src/cwe_checker_lib/src/ghidra_pcode/mod.rs

```diff
@@ -177,13 +201,415 @@ pub struct BlockSimple {
+            }
+
+            // TODO: jump targets die nicht existieren (s. IR.normalize()), leere Blöcke übernehmen.
+        }
+
+        // Special case: Block ends without jump.
+        // If all instructions are processed and current block is not new, add to blocks.
+        // TODO: Add branch here?
+        if !blk.defs.is_empty() && !blk.jmps.is_empty() {
+            blocks.push(Term {
+                tid: tid,
+                term: blk,
+            });
+        }
+
+        blocks
+    }
+}
+
+/// Translates a pcode operation and adds it to the basic block.
+///
+/// `blk` is changed by adding the corresponding operation.
+/// `tid` is changed if the pcode operation is a jump.
+/// If the operation is a jump, the block is wrapped into the tid and returned.
+/// This function returns `None`, if the operation is not a jump.
+fn add_operation_to_blk(
+    op: &PcodeOpSimple,
+    instr: &InstructionSimple,
+    mut blk: &mut Blk,
+    mut tid: &mut Tid,
+    next_instruction_address: Option<String>,
+) -> Option<Term<Blk>> {
+    match op.pcode_mnemonic {
+        // Add Def to current block.
+        PcodeOperation::ExpressionType(_) => {
+            blk.defs.append(&mut op.clone().into_ir_def(&instr.address));
+            return None;
+        }
+        PcodeOperation::JmpType(crate::pcode::JmpType::CBRANCH) => {
+            // Add conditional (not pcode relative) branch to block
+            blk.jmps.push(op.into_ir_jump(&instr.address));
+            // Add implicit branch to following instruction
+            if let Some(next_instr_address) = next_instruction_address {
+                let (finalized_blk, new_blk, new_tid) = finalize_blk_with_branch_and_setup_new_blk(
+                    &mut tid,
+                    &mut blk,
+                    format!("{}_{}", instr.address.clone(), op.pcode_index),
+                    next_instr_address,
+                );
+                // set new block and tid.
+                *blk = new_blk;
+                *tid = new_tid;
+
+                return Some(finalized_blk);
+            } else {
+                // MIPS: Delay slots for C-Jumps. Ghidra verlegt Delay ggf vor, sodas fallthrough ggf. falsch? (alter code lookup)
+                panic!("Jump target address to consecutive instruction is not provided.")
+            }
+        }
+        PcodeOperation::JmpType(jmp) => {
+            // Add potential targets
+            if matches!(
+                jmp,
+                crate::pcode::JmpType::CALLIND | crate::pcode::JmpType::BRANCHIND
+            ) {
+                if let Some(potential_targets) = &instr.potential_targets {
+                    blk.indirect_jmp_targets.append(
+                        &mut potential_targets
+                            .iter()
+                            .map(|x| Tid {
+                                id: format!("potential_target_{}", x),
+                                address: x.to_string(),
+                            })
```

Comment:
Right, I am totally with you here!
I do not understand the difference for `CALLIND` yet, since, according to P-code reference, it is semantically the same. This means they both relate to offsets and not to function names?

Reply:
For `CALLIND` the targets should be the TIDs of functions (instead of blocks) at the specific address. But we do not collect indirect call targets from Ghidra at the moment (if I remember correctly), so targets for `CALLIND` should probably lead to a `panic!`. Or we would have to build a map from addresses to function TIDs at the beginning, because we cannot guess the TIDs of functions just by their address.
But we can guess block TIDs just by their address, which is enough to handle the `BRANCHIND` case.